File: go/protein-translation/protein_translation.go
// Package protein is used to translate RNA sequences into proteins.
package protein

import (
	"errors"
	"regexp"
)

// ErrStop is used to signal FromCodon to stop processing codons.
var ErrStop = errors.New("STOP")

// ErrInvalidBase is used to signal FronCodon that an invalid codon has been encountered.
var ErrInvalidBase = errors.New("Invalid Base")

// FromRNA returns a protein sequence from an RNA sequence.
func FromRNA(rna string) ([]string, error) {
	var codons []string = []string{}
	var proteins []string = []string{}

	reStr := `([[:upper:]]{3})`

	re, e := regexp.Compile(reStr)

	if e != nil {
		panic(e)
	}

	codons = re.FindAllString(rna, -1)

	for _, codon := range codons {
		protein, e := FromCodon(codon)

		if e == ErrStop {
			break
		} else if e != nil {
			return []string{}, e
		}

		proteins = append(proteins, protein)
	}

	return proteins, nil
}

// FromCodon returns a single protein from a codon.
func FromCodon(codon string) (string, error) {
	switch codon {
	case "AUG":
		return "Methionine", nil
	case "UUU", "UUC":
		return "Phenylalanine", nil
	case "UUA", "UUG":
		return "Leucine", nil
	case "UCU", "UCC", "UCA", "UCG":
		return "Serine", nil
	case "UAU", "UAC":
		return "Tyrosine", nil
	case "UGU", "UGC":
		return "Cysteine", nil
	case "UGG":
		return "Tryptophan", nil
	case "UAA", "UAG", "UGA":
		return "", ErrStop
	default:
		return "", ErrInvalidBase
	}
}

File: ruby/protein-translation/protein_translation.rb
# frozen_string_literal: false

# https://exercism.org/tracks/ruby/exercises/protein-translation
# Protein Translation exercise
module Translation
  CODON_LENGTH = 3

  def self.to_protein(codon)
    codon2protein = {
      'AUG' => 'Methionine',
      'UUU' => 'Phenylalanine',
      'UUC' => 'Phenylalanine',
      'UUA' => 'Leucine',
      'UUG' => 'Leucine',
      'UCU' => 'Serine',
      'UCC' => 'Serine',
      'UCA' => 'Serine',
      'UCG' => 'Serine',
      'UAU' => 'Tyrosine',
      'UAC' => 'Tyrosine',
      'UGU' => 'Cysteine',
      'UGC' => 'Cysteine',
      'UGG' => 'Tryptophan',
      'UAA' => 'STOP',
      'UAG' => 'STOP',
      'UGA' => 'STOP'
    }

    raise InvalidCodonError, 'invalid codon' unless codon2protein.key?(codon)

    codon2protein[codon]
  end

  def self.of_rna(rna_sequence)
    return [] if rna_sequence.empty?

    has_stop = false
    codons = []
    rna_sequence.scan(/([A-Z]{3})/) do |match|
      codon = match[0]

      if %w[UAA UAG UGA].include?(codon)
        has_stop = true
        break
      end

      codons.push(codon)
    end

    raise InvalidCodonError, 'invalid codon' if !has_stop && rna_sequence.length.modulo(CODON_LENGTH).positive?

    codons.map { |codon| to_protein(codon) }
  end
end

# InvalidCodonError exception
class InvalidCodonError < StandardError
  attr_reader :data

  def initialize(data)
    super
    @data = data
  end
end
